File: java/src/test/java/com/kcl/PluginTest.java
package com.kcl;

import com.kcl.api.API;
import com.kcl.api.Spec.ExecProgramArgs;
import com.kcl.api.Spec.ExecProgramResult;

import java.util.Collections;

import org.junit.Assert;
import org.junit.Test;

public class PluginTest {
    @Test
    public void testExecProgramWithPlugin() throws Exception {
        API.registerPlugin("my_plugin", Collections.singletonMap("add", (args, kwArgs) -> {
            return (int) args[0] + (int) args[1];
        }));
        API api = new API();
        ExecProgramArgs exec_args = ExecProgramArgs.newBuilder().addKFilenameList("./src/test_data/plugin/plugin.k")
                .build();
        ExecProgramResult result = api.execProgram(exec_args);

        Assert.assertEquals(result.getYamlResult(), "result: 2");

        API.registerPlugin("my_plugin", Collections.singletonMap("add", (args, kwArgs) -> {
            return (int) args[20];
        }));
        api = new API();
        result = api.execProgram(exec_args);

        Assert.assertTrue("err: " + result.getErrMessage() + "yaml:" + result.getYamlResult(),
                result.getErrMessage().contains("20"));
    }
}

File: kotlin/src/test/kotlin/com/kcl/PluginTest.kt
package com.kcl

import com.kcl.api.API
import com.kcl.api.Spec.ExecProgramArgs
import com.kcl.plugin.MethodFunction
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import java.util.Collections

class PluginTest {
    @Test
    fun testExecProgramWithPlugin() {
        API.registerPlugin("my_plugin", Collections.singletonMap<String, MethodFunction>("add",
            MethodFunction { args, _ -> (args[0] as Int) + (args[1] as Int) }
        ))
        var api = API()
        val execArgs = ExecProgramArgs.newBuilder()
            .addKFilenameList("./src/test_data/plugin.k")
            .build()
        var result = api.execProgram(execArgs)
        assertEquals(result.yamlResult, "result: 2")

        API.registerPlugin("my_plugin", Collections.singletonMap<String, MethodFunction>("add",
            MethodFunction { args, _ -> args[20] }
        ))
        api = API()
        result = api.execProgram(execArgs)
        val errContains20 = result.errMessage.contains("20")
        assertTrue(
            errContains20,
            "err: ${result.errMessage} yaml:${result.yamlResult}"
        )
    }
}
